pretyping/coercion.ml

```diff
@@ -656,8 +663,12 @@ let inh_coerce_to_fail ?(use_coercions=true) flags env sigma rigidonly v v_ty ta
       in
       unify_leq_delay ~flags env sigma v'_ty target_type, v', trace
     with (Evarconv.UnableToUnify _ | Not_found) as exn ->
-      let _, info = Exninfo.capture exn in
-      Exninfo.iraise (NoCoercion,info)
+      (* 3 if none of the above works, try hook *)
+      match !hook env sigma ~flags v ~inferred:v_ty ~expected:target_type with
+      | Some (sigma, r) -> (sigma, r, ReplaceCoe r)
+      | None ->
+         let _, info = Exninfo.capture exn in
```

Comment:
```suggestion
       let _, info = Exninfo.capture exn in
      match !hook env sigma ~flags v ~inferred:v_ty ~expected:target_type with
      | Some (sigma, r) -> (sigma, r, ReplaceCoe r)
      | None ->
```
capture must be the first thing in the exception handler

Reply:
Also I'm not sure we should run the hook when `not use_coercions || (rigidonly && not (Heads.is_rigid env sigma target_type && Heads.is_rigid env sigma v_ty))` (at least when `not use_coercions`, not sure about the rigid stuff)

Reply:
I agree and it indeed seems to me that the hook is only run in the `else` branch?

Reply:
right, I misread where the `try` was